galley/pkg/config/processing/snapshotter/strategy/create.go

```diff
@@ -0,0 +1,34 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package strategy
+
+import "fmt"
+
+const (
+	debounce  = "debounce"
```

Comment:
make these public enums? Then you can take the enum value rather than a raw string in `Create` below.

Reply:
The Create method is meant to be called as part of metadata parsing:
https://github.com/ozevren/istio/blob/b5c22fb97e58921ff0b14092b248f2760becb02f/galley/pkg/config/processor/metadata/metadata.yaml#L244

Reply:
I'm not sure I see how that changes my comment. There are still a well-defined set of accepted values and making that explicit with an enum clarifies those expectations in the code.

Reply:
There needs to be at least one function that can parse a string and return something that is strongly typed. With the current version of the code, it is this method. If I make this strongly typed, then I still need to implement a string => enum, which will end up calling the enum => instance method.

Reply:
Ok, I guess I'm not seeing the call site for this function .... is it passed to some generic parsing operation where it needs to take a raw string parameter?

Reply:
The (future) call site is here: https://github.com/ozevren/istio/blob/b5c22fb97e58921ff0b14092b248f2760becb02f/galley/pkg/config/processor/build.go#L57

From this metadata object:
https://github.com/ozevren/istio/blob/b5c22fb97e58921ff0b14092b248f2760becb02f/galley/pkg/config/schema/schema.go#L87

That will be read from the Yaml above.

Reply:
Thanks for the links. I guess I'm still not seeing why an enum type can't be used here ... seems that `s.Strategy` could just be cast to the enum type, or am I missing something?

Reply:
I guess I am hung up on an enum being a value out of a discrete set of things. Arbitrarily side-casting an unknown input value to an enum doesn't seem like the right thing to do.

So, yes it is technically possible to do so. But I don't believe side-casting and querying is the right thing to do. It goes against the idea of strong typing.

Reply:
I guess I don't buy that argument. Whether you explicitly make it an enum or not is somewhat beside the point. You have a discrete set of allowed values and if it's not one of those, you return an error ... that's effectively an enum. Making it an actual string-based enum type just aids in making the code more readable.

Reply:
Whenever I see a code like this:
```
DoFoo(e MyEnum) error {
  switch e {
      case Boo: 
         Boo()
      case Bar: 
         Bar()
       default:
             return fmt.Error(...)
  }
}
```

My inclination is to change it to 
```
Do(e MyEnum) {
  switch e {
      case Boo: 
         Boo()
      case Bar: 
         Bar()
       default:
             panic(...)
             // or
             log.Errorf(...)
  }
}
```

Why? Because an enum is from a discrete set of things, and if the code is not getting the right thing, then it is better to fail fast and avoid introducing an error path due to programming error. (Which it clearly is, because either the client code is passing something bogus, or this code is out of date since a new enum is introduced). 

If I write the code in the way you're proposing, then the next guy to come along can simply do the above change, with the same expectations.

Reply:
If a developer made that change, tests would fail and it would never get through CI. 

I feel that your concern is more one of personal preference than anything. You obviously feel pretty strongly about it and it's not really worth the effort to debate it further.